### ISSUE TITLE
Readonly everything by default

### ISSUE THREAD

**Issue by safareli** [microsoft/TypeScript#42357]
# Suggestion

## 🔍 Search Terms

- Readonly by default
- Record readonly by default
- Array readonly by default https://github.com/microsoft/TypeScript/issues/32467
- Tuple readonly by default https://github.com/microsoft/TypeScript/issues/40316

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Right now default is that all record/array/tuple properties are mutable, and if you want any of them to be readonly/immutable you should add `readonly` flag or use `Readonly<...>`.  My suggestion is add a flag (or something like that) which will "flip" this - it will turn on "assume everything is read only" in TS project(or module) and add a keyword `mutable` when you want to mark something as mutable.

## 📃 Motivating Example

When using are not mutating data that much and most of the types are assumed to be immutable while very little is mutable you might accidentally mutate something or when trying to understand portion of code, which mostly uses immutable values but some are mutable, you have one option to use `readonly/Readonly..` but that code becomes quite noisy. With this flag you can turn on "readonlyByDefault" flag and everthing will be assumed to be readonly and you could mark mutable fields/values with mutable keyword. This way you would know exactly what's mutable easily and not mutate stuff accidentally.

## 💻 Use Cases

Probably 99% of react-redux projects do not mutate objects/state or use libraries for immutable structures. Also some teams where folks are using immutable values (and other functional programing practices) would benefit a lot.

---
[I proposed this initially here](https://github.com/microsoft/TypeScript/issues/32467#issuecomment-623261939) and then I noticed it had 25 👍  and suggestion to open separate proposal , which I did here.

**Comment by MartinJohns** [comment]
Please don't confuse readonly and immutable. TypeScript does not provide any mechanism to represent immutable data structures. For this reason I wouldn't add a keyword "mutable", because that kinda implies the opposite is immutable.

**Comment by safareli** [comment]
👍  Yeah, for example `writable` could be the name of the keyword. I don't have strong opinion on names whatever would be acceptable by maintainers/community is good for me.

**Comment by 00ff0000red** [comment]
So, with the flag enabled, this would error:
```ts
const x: [ number, number ] = [ 3, 9 ];
```
because the *rhs* is "readonly," but the variable doesn't accept a `readonly` tuple.

The only way to solve that would be via casting, which would beat the purpose, would it not?
```ts
const x: [ number, number ] = [ 3, 9 ] as [ number, number ];
```

**Comment by MartinJohns** [comment]
@00ff0000red

From the proposal:

> and add a keyword `mutable` when you want to mark something as mutable.

**Comment by 00ff0000red** [comment]
Oh, okay, I see. In that case it wouldn't be nearly as bad, except maybe that all of the lib typings would need to be updated, e.g. all TypedArrays are inherently mutable.

**Comment by artursvonda** [comment]
I think this would really useful feature since we'd prefer our code to be "immutable"/readonly by default and write exceptions for mutability. Currently it's very verbose.
I had basically exact same proposal in my head for this issue since it can be adopted gradually both – by each codebase and community. Utility type `MutableArray` and `MutableInterface` (or whatever) could be introduced in libraries that uses the new modifier on versions that support it and falls back to current behaviour on older versions.

**Comment by boris-kolar** [comment]
I love the idea, because I find myself writing `readonly` all the time. Immutable by default would also be very useful for compilers that compile TypeScript to wasm/native code (e.g. AssemblyScript, TypeScript Static). Also, allowing `readonly` for classes, interfaces and object types would be nice.

```ts
readonly interface Foo {
    foo: number // equivalent to: readonly foo: number
}

readonly class FooClass {
    static foo: number = 42 // equivalent to: static readonly foo: number = 42
    foo: number = 0 // equivalent to: readonly foo: number = 0
}

type Bar = readonly {
    bar: string // equivalent to: readonly bar: string
    baz: { z: number } // equivalent to: readonly baz: { readonly z: number }
}
```

**Comment by 00ff0000red** [comment]
Like @boris-kolar said, methods, prototypes, and classes should generally be immutable, by default. I'd also say that it wouldn't be too unreasonable to say that the entire global object should be readonly by default too.

I've written out readonly interfaces, and it's just painful writing readonly on every single line.

**Comment by lautarodragan** [comment]
Is this a dup of https://github.com/microsoft/TypeScript/issues/32758?

**Comment by 00ff0000red** [comment]
@lautarodragan Technicially, no, as this asks for "readonly" everything, whereas yours asks for "immutable" everything.
Otherwise, it seems so.

**Comment by isaac-weisberg** [comment]
Yeah, it makes no sense that I have to write readonly for every single interface member considering that this is my primary tool for encapsulation and access control

**Comment by benwiley4000** [comment]
> Oh, okay, I see. In that case, it wouldn't be nearly as bad, except maybe that all of the lib typings would need to be updated, e.g. all TypedArrays are inherently mutable.

I'm sure whatever implementation this takes will need to basically apply this rule only to the project and not to external modules. Any 3rd party module typings would still be "writable by default" (unless the typings file somehow declares readonly-by-default, which could/should be an option).

I can see it potentially becoming annoying to have to declare any piece of non-primitive data passed to a third party library "writable," but I imagine this annoyance would lesson over time as more libraries adopt the readonly-by-default declaration.

In the meantime there could be two workarounds: A) defensive copying (to writable) before passing (when you really don't trust the library too much), and B) `as writable` casting while passing, which would be unsafe but acceptable in cases where you know the library doesn't mutate the data.

**Comment by MaxGraey** [comment]
currently:
```ts
const tup1 = [1, 2]                             // [number, number]
const tup2: readonly [number, number] = [1, 2]  // readonly [number, number]
const tup3 = [1, 2] as const                    // readonly [number, number]
```

new:
```ts
const tup1 = [1, 2]                        // readonly [number, number]
const tup2: mut [number, number] = [1, 2]  // [number, number]
const tup3 = [1, 2] as mut                 // [number, number]
```

**Comment by imcotton** [comment]
![giphy](https://user-images.githubusercontent.com/149541/128612002-721671be-6634-4dba-ac14-4fe40267bda8.gif)

https://github.com/tc39/proposal-record-tuple

**Comment by anilanar** [comment]
This proposal, in its current form, doesn't address what happens to checking library code. For those who don't know, `tsconfig.json` flag applies to everything, including library code. What happens when a library function taking a mutable array as arg turns into a readonly array type-wise, but it actually mutates the array internally (because library source code is not type-checked, they are not included in node_modules)?

So I'm not sure how this proposal can work in a practical way without having a way to apply compiler flags (like `jsx` pragma) to subset of files.

**Comment by ivan-kleshnin** [comment]
My 5 cents. If it's not possible to implement this feature I'd personally vote for deprecation of `readonly`. It clutters code to a degree I wish it didn't exist. Given that unavoidable `as const` construct produces `readonly`, I have to either add `Writable<..>` in many places or use `readonly` everywhere.

**Comment by qpwo** [comment]
Re:

> This proposal, in its current form, doesn't address what happens to checking library code. … 
https://github.com/microsoft/TypeScript/issues/42357#issuecomment-963151858

Also considering this one: https://github.com/microsoft/TypeScript/issues/42357#issuecomment-889403488

Example I'll reference:

```ts
// myProjectFile.ts:
import { sorted } from 'somelibrary'
function smallestBiggest(supposedToBeReadonlyNums: number[]): [number, number] {
    const sortedArr = sorted(supposedToBeReadonlyNums) // line 4
    return [sortedArr[0], sortedArr.at(-1)]
}

// somelibrary/sorted.ts:
function sorted(arr: number[]): number[] {
    arr.sort((x, y) => x - y) // mutation!
    return arr
}
```

Not knowing anything about typescript's internals, (but as a heavy user) here's a few options  – mix of things the compiler could do or the user could do –  not sure if they're all feasible:

1. Allow compiler flags to subset of files as you stated. Line 4 then throws no error? Or you put `as writeable` before the close paren? Not sure I understand @anilanar 's comment.
2. Have a `declare global` block in your project somewhere stating which functions are `writeable` and which are `readonly`. This block could be moved to the lib typings with a PR eventually, and copy-pasted around meanwhile.
3. Um sorry for asking this but how impossible is automatic detection of mutating operations ? … Like, the compiler can already detect violations of `readonly` so can't it automatically infer in all library code when a function violates it and the whole question is moot? This currently doesn't 100% work with untyped JS – or work at all?
4. Again my knowledge is lacking here but would there be an easy way to disable the "Readonly everything by default" setting for a single file? Then you could wrap all the library functions you need without too much clutter. Similar to option 2.
5. Objects/arrays originating from library code would be writeable but data from user factories would be readonly, right? Does this already avoid a lot of the annoying cases of casting stuff to/from readonly all the time? (Really uncertain about this one but throwing it in for the sake of variety.)

**Comment by TheUnlocked** [comment]
Rather than have `readonly` by default (which seems dangerous for interacting with libraries), I think it would make more sense for the flag to just disallow unannotated array/object types (and maybe a separate flag for whether literals should be writable/readonly, though that might not be worth it).

**Comment by MartinJohns** [comment]
@anilanar Please don't confuse read-only with immutable.

**Comment by anilanar** [comment]
@MartinJohns Perhaps this issue and its dual #32758 require some enlightenment about that. Does true immutability exist in TS? If it doesn't, don't we refer to `readonly` fields and `ReadonlyArray`s when we say immutable in this context?

**Comment by anilanar** [comment]
Even the type error itself mentions mutability: `The type 'readonly number[]' is 'readonly' and cannot be assigned to the mutable type 'number[]'.`

**Comment by MartinJohns** [comment]
@anilanar Immutability can't be represented in TypeScripts type system as of today. Read-only only means it's read only via that interface. A good example is that mutable objects can be implicitly assigned to the read-only versions. The object is read only, but it's not immutable, it can still be mutated just fine. (side note: the read-only version can be passed implicitly to the mutable version as well.)

**Comment by soplwang** [comment]
"Readonly everything by default" is just like Rust does and compliant with _"Principle of least privilege"_ that's awesome.

Every language that not readonly by default should consider unsafe and should obsoleted.

**Comment by soplwang** [comment]
> "Readonly everything by default" is just like Rust does and compliant with _"Principle of least privilege"_ that's awesome.
> 
> Every language that not readonly by default should consider unsafe and should obsoleted.

Seems `"use safety"` directive like solution more appropriate for future JS standard and TS.

**Comment by mjljm** [comment]
Hi! Has any progress been made on that topic? I am also very interested in a readonly bu default typescript flag in tsconfig.json that would apply only to the files covered by this tsconfig.json.

**Comment by Tungetyt** [comment]
Please let it happen.

**Comment by rhaksw** [comment]
@MartinJohns @anilanar to move this forward, is correct that @safareli would need to modify the proposal to account for the impact on library code, as @anilanar mentioned here in [Jan 2021](https://github.com/microsoft/TypeScript/issues/32467#issuecomment-760645387), and [here in Nov 2021](https://github.com/microsoft/TypeScript/issues/42357#issuecomment-963151858)?

CC @kachkaev for input on why the current proposal is undesirable.

**Comment by rhaksw** [comment]
Thank you for that explanation, @anilanar. Speaking as an application developer, I do not use "readonly" (either the flag, `Readonly<...>`, or `const ... as const`, not because it makes libraries harder to use, but rather because it requires me to alter all of my variable declarations. If a library required me to mutate its objects in my own code, then I simply would not use it.

This proposal interested me because it would not require me to alter all variable declarations. I understand that if I adopted the proposed "readonlyByDefault", then that may limit the libraries available to me. But then, unless I am misunderstanding something here, I would also consider those to be poorly written libraries that I shouldn't be using anyway.

**Comment by rhaksw** [comment]
Thinking about this again, maybe an ESLint plugin such as [eslint-plugin-functional](https://github.com/eslint-functional/eslint-plugin-functional?tab=readme-ov-file#eslint-plugin-functional) would be more practical than trying to change TypeScript.

**Comment by rhaksw** [comment]
Yes, I meant the rule in your second link, `functional/immutable-data`. Thanks!

**Comment by andyearnshaw** [comment]
Originally I posted this comment in https://github.com/microsoft/TypeScript/issues/7770, as I found that issue before this one. However, I think it's valid as an example here too.

--- 

I frequently run into the same scenario over and over in React where I see people reaching for memoization unnecessarily. It usually ends up looking like this:

```typescript
function MyComponent({ config }: Props) {
  // Prevent recreating the default array on every call to MyComponent
  const someArray = useMemo(() => config?.someArray ?? [], [config]);

  return <SomeOtherComponent someArray={someArray};
}
```

This is a bit of a trigger for me when I'm reviewing pull requests, but I see it several times per week. I'll usually ask the author to change it to use an empty array outside the scope of the component:

```typescript
// Frozen because any function/component that modifies it would be disastrous.
const DEFAULT_ARRAY = Object.freeze([]);

function MyComponent({ config }: Props) {
  // Prevent recreating the default array on every call to MyComponent
  const someArray = config?.someArray ?? DEFAULT_ARRAY as Props['someArray'];

  return <SomeOtherComponent someArray={someArray} />;
}
```

However, this might produce a potential refactoring hazard, because `SomeOtherComponent` is not expecting a `readonly` array. Usually, I end up rewriting it to be `as unknown as Props['someArray']`, one of the few instances of `as unknown as` I will tolerate in production code.

What I would really like to see is TypeScript move towards read-only types by default, with a `writable` state to counter the `readonly` state.

```typescript
// The Foo type has the readonly modifier.
type Foo = {
  someProp: number
};

// Incorrect
function mySideEffectFunction(
  foo: Foo,
) {
  foo.someProp = 2; // type error, cannot modify readonly object

  // ...
}

// Correct
function mySideEffectFunction(
  foo: Writable<Foo>,
) {
  foo.someProp = 2;

  // ...
}
```
